### PR TITLE
Feature/bca/improve leave space

### LINKED ISSUE #3692: M9.1 Spaces | Allow to also leave rooms when leaving a space
https://www.figma.com/file/Sx8vQd8cvrrjjIDrBz9uLw/%5BBeta%5D-Leave-a-space?node-id=78%3A28693
image

image

~3d

### PR DESCRIPTION
**Pull request by BillCarsonFr**
Fixes #3692
M9.1 Spaces | Allow to also leave rooms when leaving a space 

When leaving you have 3 choices, leave all rooms of this space , leave none, or select which one to leave.
Also displays a warning message if you won't be able to re-join or if you are the last admin ...

image

### CODE REVIEW

File: vector/src/main/res/layout/bottom_sheet_leave_space.xml [hunk 1]
```diff
@@ -0,0 +1,105 @@
...
+    <TextView
+        android:id="@+id/inlineErrorText"
+        style="@style/Widget.Vector.TextView.Body"
+        android:layout_width="match_parent"
+        android:layout_height="wrap_content"
+        android:layout_marginStart="@dimen/layout_horizontal_margin"
+        android:layout_marginTop="4dp"
+        android:layout_marginEnd="@dimen/layout_horizontal_margin"
+        android:textColor="?vctr_content_secondary"
+        android:visibility="gone" />
```
Comment (bmarty) [inline vector/src/main/res/layout/bottom_sheet_leave_space.xml:63]: add tools:visibility="visible" ?

File: vector/src/main/res/values/strings.xml [hunk 2]
```diff
@@ -3416,11 +3416,15 @@
     <string name="space_explore_activity_title">Explore rooms</string>
     <string name="space_add_child_title">Add rooms</string>
     <string name="leave_space">Leave Space</string>
-    <string name="space_leave_prompt_msg">Are you sure you want to leave the space?</string>
+    <string name="space_leave_prompt_msg">Are you sure you want to leave %s?</string>
```
Comment (bmarty) [inline vector/src/main/res/values/strings.xml:3419]: The existing translations if any will fail :/

File: vector/src/main/res/values/strings.xml [hunk 3]
```diff
@@ -3416,11 +3416,15 @@
...
     <string name="space_leave_prompt_msg_only_you">You are the only person here. If you leave, no one will be able to join in the future, including you.</string>
-    <string name="space_leave_prompt_msg_private">This space is not public. You will not be able to rejoin without an invite.</string>
-    <string name="space_leave_prompt_msg_as_admin">You are admin of this space, ensure that you have transferred admin right to another member before leaving.</string>
-
+    <string name="space_leave_prompt_msg_private">You won\'t be able to rejoin unless you are re-invited.</string>
+    <string name="space_leave_prompt_msg_as_admin">You\'re the only admin of this space. Leaving it will mean no one has control over it.</string>
+    <string name="leave_all_rooms_and_spaces">Leave all rooms and spaces</string>
+    <string name="you_will_leave_all_in">You will leave all rooms and spaces in %s.</string>
+    <string name="dont_leave_any">Don’t leave any rooms and spaces</string>
+    <string name="leave_specific_ones">Leave specific rooms and spaces</string>
```
Comment (bmarty) [inline vector/src/main/res/values/strings.xml:3426]: Maybe end the string with "..." because there will be another step.

**Review by bmarty — CHANGES_REQUESTED** [review]
The UX is a bit weird when selecting "Leave specific rooms and spaces", but it's working.

On the bottom sheet, it could be better to align horizontally "Leave Space"  and "Cancel". But I would remove the "Cancel" anyway, it's a bit useless, and aligne "Leave space" to the right. Also the text color is not correct in dark theme (it's black and should be white. Bg is red which is OK)

On the Pick room Activity, It should be nice to have a option to select / unselect all rooms.

When pressing back from the Activity (I would also remove the cancel here), we should stay on the bottom sheet with the 3rd radio button check and would only be able to leave the space from here (and not from the Activity).

UX is a bit too bad to be mergeable.

File: vector/src/main/java/im/vector/app/features/spaces/SpaceMenuViewModel.kt [hunk 4]
```diff
@@ -0,0 +1,162 @@
...
+                                ?.map { powerLevelsHelper.getUserRole(it) }
+                                ?.count { it is Role.Admin }
+                                ?: 0
+                        val isLastAdmin = isAdmin && otherAdminCount == 0
+
+                        setState {
+                            copy(
+                                    canEditSettings = canChangeAvatar || canChangeName || canChangeTopic,
+                                    canInvite = canInvite,
+                                    canAddChild = canAddChild,
```
[the remark this answers is not in the document]
Comment (BillCarsonFr) [reply vector/src/main/java/im/vector/app/features/spaces/SpaceMenuViewModel.kt:111]: It's used by the parent bottom sheet that is using the same view model (to show invite/manage/addrooms buttons)

File: vector/src/main/java/im/vector/app/features/spaces/leave/SpaceLeaveAdvancedViewModel.kt [hunk 5]
```diff
@@ -0,0 +1,135 @@
...
+            is SpaceLeaveAdvanceViewAction.ToggleSelection -> {
+                val existing = state.selectedRooms.toMutableList()
+                if (existing.contains(action.roomId)) {
+                    existing.remove(action.roomId)
+                } else {
+                    existing.add(action.roomId)
+                }
+                setState {
+                    copy(
+                            selectedRooms = existing.toList()
```
[the remark this answers is not in the document]
Comment (BillCarsonFr) [reply vector/src/main/java/im/vector/app/features/spaces/leave/SpaceLeaveAdvancedViewModel.kt:59]: I thought it was returning a immutable view of the list, but it's not. Removed

File: vector/src/main/java/im/vector/app/features/spaces/leave/SpaceLeaveAdvancedViewModel.kt [hunk 6]
```diff
@@ -0,0 +1,135 @@
...
+                setState { copy(leaveState = Loading()) }
+                viewModelScope.launch {
+                    try {
+                        try {
+                            state.selectedRooms.forEach {
+                                session.getRoom(it)?.leave(null)
+                            }
+                        } catch (failure: Throwable) {
+                            // silently ignore?
+                            Timber.e(failure, "Fail to leave sub rooms/spaces")
```
[the remark this answers is not in the document]
Comment (BillCarsonFr) [reply vector/src/main/java/im/vector/app/features/spaces/leave/SpaceLeaveAdvancedViewModel.kt:76]: good catch

File: vector/src/main/res/layout/bottom_sheet_leave_space.xml [hunk 7]
```diff
@@ -0,0 +1,105 @@
+<?xml version="1.0" encoding="utf-8"?>
+<LinearLayout xmlns:android="http://schemas.android.com/apk/res/android"
+    xmlns:tools="http://schemas.android.com/tools"
+    android:layout_width="match_parent"
+    android:layout_height="wrap_content"
+    android:background="?colorSurface"
```
[the remark this answers is not in the document]
Comment (BillCarsonFr) [reply vector/src/main/res/layout/bottom_sheet_leave_space.xml:6]: done

File: vector/src/main/res/values/strings.xml (hunk 3 above)
Comment (BillCarsonFr) [reply vector/src/main/res/values/strings.xml:3426]: ok

File: vector/src/main/res/layout/bottom_sheet_leave_space.xml (hunk 1 above)
Comment (BillCarsonFr) [reply vector/src/main/res/layout/bottom_sheet_leave_space.xml:63]: done

File: vector/src/main/java/im/vector/app/features/spaces/leave/SpaceLeaveAdvancedFragment.kt [hunk 8]
```diff
@@ -0,0 +1,85 @@
...
+    override fun invalidate() = withState(viewModel) { state ->
+        super.invalidate()
+        controller.setData(state)
+    }
+
+    override fun onItemSelected(roomSummary: RoomSummary) {
+        viewModel.handle(SpaceLeaveAdvanceViewAction.ToggleSelection(roomSummary.roomId))
+    }
+
+    override fun onBackPressed(toolbarButton: Boolean): Boolean {
```
[the remark this answers is not in the document]
Comment (BillCarsonFr) [reply vector/src/main/java/im/vector/app/features/spaces/leave/SpaceLeaveAdvancedFragment.kt:81]: tested, and not needed, i removed it. Thx

File: vector/src/main/java/im/vector/app/features/spaces/leave/SpaceLeaveAdvancedActivity.kt [hunk 9]
```diff
@@ -0,0 +1,127 @@
...
+                is Loading -> {
+                    showWaitingView()
+                }
+                is Success -> {
+                    setResult(RESULT_OK)
+                    finish()
+                }
+                is Fail    -> {
+                    hideWaitingView()
+                    showSnackbar(errorFormatter.toHumanReadable(state.leaveState.error))
```
[the remark this answers is not in the document]
Comment (BillCarsonFr) [reply vector/src/main/java/im/vector/app/features/spaces/leave/SpaceLeaveAdvancedActivity.kt:107]: done, it's using a dialog now

File: vector/src/main/java/im/vector/app/features/spaces/leave/SpaceLeaveAdvancedActivity.kt [hunk 10]
```diff
@@ -0,0 +1,127 @@
...
+
+    override fun showWaitingView(text: String?) {
+        hideKeyboard()
+        views.waitingView.waitingStatusText.isGone = views.waitingView.waitingStatusText.text.isNullOrBlank()
+        super.showWaitingView(text)
+    }
+
+    override fun hideWaitingView() {
+        views.waitingView.waitingStatusText.text = null
+        views.waitingView.waitingStatusText.isGone = true
```
[the remark this answers is not in the document]
Comment (BillCarsonFr) [reply vector/src/main/java/im/vector/app/features/spaces/leave/SpaceLeaveAdvancedActivity.kt:67]: done

File: vector/src/main/res/values/strings.xml (hunk 2 above)
Comment (BillCarsonFr) [reply vector/src/main/res/values/strings.xml:3419]: deleted the old one, and added a new one

File: vector/src/main/java/im/vector/app/features/spaces/SpaceMenuViewAction.kt [hunk 11]
```diff
@@ -0,0 +1,26 @@
...
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package im.vector.app.features.spaces
+
+import im.vector.app.core.platform.VectorViewModelAction
+
+sealed class SpaceMenuViewAction : VectorViewModelAction {
```
[the remark this answers is not in the document]
Comment (bmarty) [reply vector/src/main/java/im/vector/app/features/spaces/SpaceMenuViewAction.kt]: OK

File: vector/src/main/java/im/vector/app/features/spaces/SpaceMenuViewModel.kt (hunk 4 above)
[the remark this answers is not in the document]
Comment (bmarty) [reply vector/src/main/java/im/vector/app/features/spaces/SpaceMenuViewModel.kt:113]: OK

**Review by bmarty — CHANGES_REQUESTED** [review]
Thanks for the update. One remark about translations, sorry :/

File: vector/src/main/res/values-ar/strings.xml [hunk 12]
```diff
@@ -1165,7 +1165,6 @@
     <string name="space_explore_activity_title">استكشِف الغُرف</string>
     <string name="space_add_child_title">أضف غُرف</string>
     <string name="leave_space">غادر المساحة</string>
-    <string name="space_leave_prompt_msg">هل أنت متأكد أنك تريد مغادرة المساحة؟</string>
```
[the remark this answers is not in the document]
Comment (BillCarsonFr) [reply vector/src/main/res/values-ar/strings.xml:1168]: Dp,e

**Pull request merged**